### PR TITLE
feat: support explicit split inputs and nn epoch overrides

### PR DESCRIPTION
**Pull request by marcellodebernardi**
This PR adds explicit validation/test split support in `plexe` data preparation and updates entrypoint epoch overrides to `nn_default_epochs`/`nn_max_epochs`, while preserving backward compatibility for existing dataset input usage. The aim is to let callers provide train/val/test inputs directly (or partially) with predictable split generation behavior and cleaner neural-network epoch controls from the CLI/function entrypoint.

## Testing
- `poetry run pytest tests/unit/agents/test_dataset_splitter_prompt.py tests/unit/test_main_dataset_inputs.py tests/unit/workflow/test_prepare_data_explicit_splits.py tests/unit/workflow/test_column_exclusion.py`
- `poetry run ruff check plexe/main.py plexe/workflow.py plexe/agents/dataset_splitter.py tests/unit/test_main_dataset_inputs.py tests/unit/workflow/test_prepare_data_explicit_splits.py tests/unit/agents/test_dataset_splitter_prompt.py tests/integration/conftest.py`

### CODE REVIEW

File: plexe/workflow.py [hunk 1]
```diff
@@ -861,91 +925,105 @@ def prepare_data(
...
+                    "Splitter returned an unexpected third split while generating test-only split; "
+                    "using returned test URI"
                 )
-                split_ratios = {"train": 0.7, "val": 0.15, "test": 0.15}
-
-            logger.info("Creating train/val/test splits from single dataset (final evaluation enabled)")
+                test_uri = split_test_uri
+            else:
+                # In this path the 2-way splitter "val" output is intentionally repurposed as test split.
+                test_uri = generated_test_uri
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/workflow.py:1007]: Kept this branch aligned with the current splitter contract in this PR scope: when only validation is supplied and a test split is requested, the 2-way splitter output is intentionally treated as the auxiliary test artifact. I am keeping output routing unchanged here to avoid introducing a second split-output contract in the same change set.

[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/workflow.py:1007]: Addressed in the latest push: the val-only + generate-test branch now runs the splitter in a dedicated `generated` subdirectory under splits output, so explicit `val.parquet` is not overwritten. Added a regression assertion in `test_prepare_data_generates_missing_test_when_only_val_is_provided` to enforce this output routing.

File: plexe/main.py [hunk 2]
```diff
@@ -36,17 +37,20 @@
...
     max_iterations: int = 10,
     global_seed: int | None = None,
     work_dir: Path = Path("/tmp/model_builder_v2"),
+    train_dataset_uri: str | None = None,
+    val_dataset_uri: str | None = None,
     test_dataset_uri: str | None = None,
     enable_final_evaluation: bool = False,
-    max_epochs: int | None = None,
+    nn_default_epochs: int | None = None,
+    nn_max_epochs: int | None = None,
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/main.py:53]: Kept the entrypoint epoch override surface focused on `nn_default_epochs` and `nn_max_epochs` in this change set. The backward-compatibility path retained here is the deprecated `data_refs` dataset input fallback.

File: plexe/main.py [hunk 3]
```diff
@@ -110,10 +117,13 @@ def main(
...
         config.spark_mode = spark_mode
-        if max_epochs:
-            config.nn_max_epochs = max_epochs
-            if config.nn_default_epochs > config.nn_max_epochs:
-                config.nn_default_epochs = config.nn_max_epochs
+        epoch_overrides = {}
+        if nn_default_epochs is not None:
+            epoch_overrides["nn_default_epochs"] = nn_default_epochs
+        if nn_max_epochs is not None:
+            epoch_overrides["nn_max_epochs"] = nn_max_epochs
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/main.py:124]: Keeping fail-fast validation behavior for inconsistent epoch overrides in this path, so invalid override combinations are surfaced immediately instead of being silently adjusted.

File: plexe/main.py [hunk 4]
```diff
@@ -294,7 +364,16 @@ def _on_checkpoint(phase_name, checkpoint_path, work_dir):
...
+    parser.add_argument(
+        "--nn-default-epochs",
+        type=int,
+        help="Override default epochs for neural network training (Keras, PyTorch)",
+    )
+    parser.add_argument(
+        "--nn-max-epochs",
+        type=int,
+        help="Override max epochs cap for neural network training (Keras, PyTorch)",
+    )
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/main.py:382]: The CLI in this PR intentionally uses explicit neural-network epoch flags (`--nn-default-epochs`, `--nn-max-epochs`) to keep the override semantics unambiguous.

File: plexe/workflow.py [hunk 5]
```diff
@@ -861,91 +925,105 @@ def prepare_data(
...
+
+    if provided_val and provided_test:
+        logger.info("Using provided train/val/test splits without splitter")
+    elif provided_val and not provided_test:
+        if generate_test_set:
+            logger.info("Validation split provided - generating missing test split from training dataset")
+            split_ratios = get_three_way_split_ratios()
+            test_ratio = split_ratios.get("test", 0.15)
+            split_ratios = {"train": max(1.0 - test_ratio, 0.01), "val": test_ratio}
+            train_uri, generated_test_uri, split_test_uri = run_split(split_ratios)
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/workflow.py]: Fixed in the latest commit: this branch now calls the splitter with a dedicated generated output directory (`.../splits/generated`) when explicit val is provided and only test must be generated, which avoids overwriting materialized explicit validation data.

File: plexe/workflow.py [hunk 6]
```diff
@@ -974,36 +1052,8 @@ def prepare_data(
...
-
-        extra_in_test = test_features - train_features
-        if extra_in_test:
-            logger.warning(
-                f"Test dataset has {len(extra_in_test)} extra features not in training data (will be ignored)"
-            )
-
-        logger.info("✓ Test dataset schema validation complete")
+    # TODO(explicit-split-schema-validation): Validate explicit val/test schema compatibility
+    # against training schema (target + feature alignment) in a dedicated follow-up PR.
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/workflow.py:1062]: Schema alignment checks are intentionally left as the dedicated follow-up tracked by the TODO in this PR, to keep this change focused on explicit split routing and input handling behavior.

File: plexe/workflow.py [hunk 7]
```diff
@@ -107,6 +107,7 @@ def _apply_allowed_model_types_on_resume(context: BuildContext, config: Config,
 def build_model(
     spark: SparkSession,
     train_dataset_uri: str,
+    val_dataset_uri: str | None,
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/workflow.py:110]: Kept this signature explicit in the workflow layer and updated call sites accordingly in this change, so the new optional validation input is always passed intentionally by callers.

File: plexe/workflow.py [hunk 8]
```diff
@@ -848,9 +882,39 @@ def analyze_data(
...
+    split_df.write.mode("overwrite").parquet(output_uri)
+    logger.info(f"Copied {split_name} dataset to: {output_uri}")
+    return output_uri
+
+
 def prepare_data(
     spark: SparkSession,
     training_dataset_uri: str,
+    val_dataset_uri: str | None,
     test_dataset_uri: str | None,
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/workflow.py:918]: Similarly kept `prepare_data` explicit at this layer, with all internal call sites updated to pass validation/test dataset inputs directly for clarity of split-resolution behavior.

**Comment by marcellodebernardi** [comment]
@greptileai please review again with latest changes

File: plexe/main.py [hunk 9]
```diff
@@ -110,10 +117,13 @@ def main(
...
-            config.nn_max_epochs = max_epochs
-            if config.nn_default_epochs > config.nn_max_epochs:
-                config.nn_default_epochs = config.nn_max_epochs
+        epoch_overrides = {}
+        if nn_default_epochs is not None:
+            epoch_overrides["nn_default_epochs"] = nn_default_epochs
+        if nn_max_epochs is not None:
+            epoch_overrides["nn_max_epochs"] = nn_max_epochs
+        if epoch_overrides:
+            config = config.__class__.model_validate(config.model_dump() | epoch_overrides)
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/main.py]: Applied in the latest push: when only `nn_max_epochs` is overridden, `nn_default_epochs` is now clamped to the new cap before `model_validate`, preserving the previous max-only behavior. Also added `test_main_nn_max_epochs_override_clamps_default_when_only_cap_is_set` to lock this in.

**Comment by marcellodebernardi** [comment]
@greptileai please review again with latest changes

File: plexe/main.py [hunk 10]
```diff
@@ -36,17 +37,20 @@
...
     global_seed: int | None = None,
     work_dir: Path = Path("/tmp/model_builder_v2"),
+    train_dataset_uri: str | None = None,
+    val_dataset_uri: str | None = None,
     test_dataset_uri: str | None = None,
     enable_final_evaluation: bool = False,
-    max_epochs: int | None = None,
+    nn_default_epochs: int | None = None,
+    nn_max_epochs: int | None = None,
     allowed_model_types: list[str] | None = None,
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/main.py:54]: This change intentionally moves the epoch override interface to `nn_default_epochs`/`nn_max_epochs` and keeps compatibility handling scoped to dataset input migration (`data_refs`). Existing callers should switch to the new epoch override parameters in this release.

File: plexe/main.py [hunk 11]
```diff
@@ -155,21 +171,76 @@ def main(
...
+        csv_options = {"sep": config.csv_delimiter, "header": config.csv_header}
         train_dataset_uri, input_format = normalizer.normalize(
-            input_uri=input_uri, output_uri=normalized_output, read_options=csv_options
+            input_uri=resolved_train_input_uri, output_uri=normalized_output, read_options=csv_options
         )
         input_format = input_format.value
 
+        # Optional explicit split datasets are ignored in retraining mode
+        if is_retrain and (val_dataset_uri or test_dataset_uri):
+            logger.warning("val_dataset_uri/test_dataset_uri are ignored in retraining mode")
```
[the remark this answers is not in the document]
Comment (marcellodebernardi) [reply plexe/main.py:210]: Using the effective normalized training URI here is intentional so auxiliary normalized inputs are colocated with the same integration-selected artifact root used for the run context.
